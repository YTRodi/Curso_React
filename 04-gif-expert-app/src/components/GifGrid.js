import React, { useEffect, useState } from 'react'
import GifGridItem from './GifGridItem';

const GifGrid = ( { category } ) => {
  
    const [images, setImages] = useState([]);

    // Me permite poder ejecutar código de manera condicional, es decir,
    // solamente quiero que se ejecute esa condición cuando el componente es RENDERIZADO por PRIMERA vez.
    // ( Analogía al patrón de diseño Singleton )
    //     1er param = función a ejecutar
    //     2do param = [] de dependencias
    useEffect( () => {

        getGifs();

    }, [])

    // Función para consumir los datos de la Fetch API
    const getGifs = async() => {

        // endpoint
        const url = 'https://api.giphy.com/v1/gifs/search?q=Tanjiro&limit=10&api_key=Nt4ACMs0TBmsydLDNeElO00qxCJaqUTp'; 
        const res = await fetch( url );
        // const data = await res.json();
        const { data } = await res.json();
        
        // No me interesan todos los datos que me proporciona la data, solo algunos por eso mapeo...
        const onlyGifs = data.map( img => {

            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url // images? = Preguntamos que si viene una imagen, vamos a utilizar esta prop
            }
            
            // return { id: img.id, title: img.title, url: img.images?.downsized_medium.url }  // Destructuring
        });

        console.log( onlyGifs );
        setImages( onlyGifs );

    }

    return (
        <>
            <h3>{ category }</h3>
            <div className="card-grid">
                <ol>
                    {
                        // #1
                        // images.map( oneImg => 

                        //     <li key={ oneImg.id }>{ oneImg.title }</li>
                        // )

                        // #2 destructuring
                        // images.map( ( { id, title } ) => 

                        //     <li key={ id }>{ title }</li>
                        // )
                        
                        // #2 con el componente nuevo GifGridItem
                        images.map( img  => (

                            <GifGridItem 
                                key={ img.id }
                                { ...img } // Estoy enviando cada una de las props de las img como una prop independiente.
                                // img={ img } 
                            />

                        ))
                    }
                </ol>
            </div>
        </>
    )
}

export default GifGrid;
