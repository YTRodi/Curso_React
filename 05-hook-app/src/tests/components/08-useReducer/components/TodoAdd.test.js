import { shallow } from "enzyme";
import { TodoAdd } from "../../../../components/08-useReducer/components/TodoAdd";

describe('Pruebas en <TodoAdd />', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow( 
        <TodoAdd 
            handleAddTodo={ handleAddTodo }
        /> 
    );
    
    test('Debe de mostrar el componente correctamente', () => { 
        expect( wrapper ).toMatchSnapshot();
    });

    test('NO debe de llamar handleAddTodo', () => {
        
        // description.length <= 1
        const formSubit = wrapper.find( 'form' ).prop( 'onSubmit' ); // Hace ref a una función.
        
        formSubit( { preventDefault(){} } );

        expect( handleAddTodo ).toHaveBeenCalledTimes( 0 );
        
    });

    test('Debe de llamar la función handleAddTodo', () => {
        
        // Con un argumento
        const value = 'Aprender React';
        wrapper.find( 'input' ).simulate( 'change', {
            target: {
                value,
                name: 'description'
            }
        });

        const formSubit = wrapper.find( 'form' ).prop( 'onSubmit' ); // Hace ref a una función.
        formSubit( { preventDefault(){} } );
        
        expect( handleAddTodo ).toHaveBeenCalledTimes( 1 );
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any( Object ) ); // {  } standard
        expect( handleAddTodo ).toHaveBeenCalledWith( {
            id: expect.any(Number), // No me importa que id sea, sólo me importa que sea un número
            desc: value,
            done: false,
        } );

        expect( wrapper.find( 'input' ).prop( 'value' ) ).toBe( '' );

    });

});