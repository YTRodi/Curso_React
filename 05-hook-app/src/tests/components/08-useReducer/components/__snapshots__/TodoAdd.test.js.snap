// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd /> Debe de mostrar el componente correctamente 1`] = `
<Fragment>
  <h4>
    add TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="false"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Your new todo..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
